packages/web/src/modules/zoom/components/zoom-provider.tsx

```diff
@@ -0,0 +1,149 @@
+import React, { useEffect, useCallback, useState, ChangeEvent } from 'react';
+import { hooks, ui } from '@application';
+import { Components } from '@types';
+
+import { zoomAuthorize, generateSignature } from '../requests';
+import { ZoomContext, ZoomContextType, Role } from '../context';
+import { authorizeZoom } from '../services';
+import ZoomActivityView from './zoom-activity-view';
+
+const { useApi, useQuery } = hooks;
+const { Button, Container, Input, Row, Col, Spinner } = ui;
+
+const ZoomProvider: Components['ZoomProvider'] = ({ redirectUri }) => {
+  const [zoomContextState, setZoomContextState] = useState<ZoomContextType>({
+    userSignature: '',
+    hostUserZakToken: null,
+    meetingNumber: '',
+    username: '',
+    password: '',
+    role: null,
+  });
+
+  const { code } = useQuery<{ code?: string; path?: string }>();
+  const zoomAuthorizeApi = useApi(zoomAuthorize);
+  const zoomSignatureApi = useApi(generateSignature);
+
+  useEffect(() => {
+    if (
+      code &&
+      zoomContextState.role !== null &&
+      !zoomSignatureApi.response &&
+      !zoomSignatureApi.loading
+    ) {
+      if (zoomContextState.role === Role.Host) {
+        zoomAuthorizeApi.fetch({ code, redirectUri });
```

Comment:
You're doing 2 parallel request when host?

Reply:
Yes, one for the Zak token one for the signature.
Should be better to have unified?

Reply:
What do you mean by unified?

It doesn't seem logical to have 2 parallel requests, this indicates they're not related. If so, what is the purpose of first one?

Furthermore, it looks very odd that they have the same exact condition when those should be made? I'd say by feeling that ZAK has a different lifecycle.

Reply:
We can do everything in one request and handle the conditions on the backend, but I'm not sure that's a good practice...

`Signature` is always needed (both host and guest) and `auth` is needed only for host to get the ZAK token. Both things are provided to Zoom upon starting a call. That's why in the case of a host, both are fetched at the same time